### PR TITLE
KAFKA-14133: Move consumer mock in TaskManagerTest to Mockito - part 2

### PR DESCRIPTION
**Pull request by clolov**
The previous pull request in this series was https://github.com/apache/kafka/pull/15112.

This pull request continues the migration of the consumer mock in TaskManagerTest test by test for easier reviews.

I envision there will be at least 1 more pull request to clean things up. For example, all calls to `taskManager.setMainConsumer` should be removed.

### CODE REVIEW

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java [hunk 1]
```diff
@@ -2358,6 +2354,7 @@ public void shouldCloseActiveUnassignedSuspendedTasksWhenClosingRevokedTasks() {
         assertThat(taskManager.activeTaskMap(), Matchers.anEmptyMap());
         assertThat(taskManager.standbyTaskMap(), Matchers.anEmptyMap());
         Mockito.verify(activeTaskCreator).closeAndRemoveTaskProducerIfNeeded(taskId00);
+        Mockito.verify(mockitoConsumer).resume(assignment);
```
Comment (clolov) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:2357]: There are quite a few tests where this verification did not hold true. By spot-checking this in EasyMock the tests appeared to pass without this change there as well.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java [hunk 2]
```diff
@@ -4821,8 +4911,10 @@ private Map<TaskId, StateMachineTask> handleAssignment(final Map<TaskId, Set<Top
         when(standbyTaskCreator.createTasks(standbyAssignment)).thenReturn(standbyTasks);
         when(activeTaskCreator.createTasks(any(), Mockito.eq(allActiveTasksAssignment))).thenReturn(allActiveTasks);
 
-        expectRestoreToBeCompleted(consumer);
-        replay(consumer);
+        final Set<TopicPartition> assignment = singleton(new TopicPartition("assignment", 0));
+        lenient().when(mockitoConsumer.assignment()).thenReturn(assignment);
```
Comment (clolov) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:4811]: `lenient()` was needed here because this method is a dependency on quite a few tests some of which do not actually call this. The tests which do not call it are
* `shouldClassifyExistingTasksWithoutStateUpdater`
* `shouldComputeOffsetSumForNonRunningActiveTask`
* `shouldSkipUnknownOffsetsWhenComputingOffsetSum`
* `shouldReportLatestOffsetAsOffsetSumForRunningTask`

**Comment by clolov** [comment]
Heya @lucasbru @divijvaidya and @cadonna - a polite call for review 😊! There are only 2 classes left to move to Mockito!

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java [hunk 3]
```diff
@@ -4821,8 +4911,10 @@ private Map<TaskId, StateMachineTask> handleAssignment(final Map<TaskId, Set<Top
...
+
+        taskManager.setMainConsumer(mockitoConsumer);
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:4813]: Why did you not modify this call in `setUpTaskManager()`?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java [hunk 4]
```diff
@@ -2016,13 +2015,14 @@ public void shouldReleaseLockForUnassignedTasksAfterRebalance() throws Exception
...
         handleAssignment(taskId00Assignment, taskId01Assignment, emptyMap());
-        reset(consumer);
-        expectConsumerAssignmentPaused(consumer);
-        replay(consumer);
 
         taskManager.handleRebalanceComplete();
         assertThat(taskManager.lockedTaskDirectories(), is(mkSet(taskId00, taskId01)));
         verify(stateDirectory);
+
+        final Set<TopicPartition> assignment = singleton(new TopicPartition("assignment", 0));
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java]: This assignment is duplicated a couple of times. Could you try to avoid that duplication?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java [hunk 5]
```diff
@@ -2210,6 +2210,9 @@ public void shouldComputeOffsetSumForStandbyTask() throws Exception {
         restoringTask.setChangelogOffsets(changelogOffsets);
 
         assertThat(taskManager.getTaskOffsetSums(), is(expectedOffsetSums));
+
+        final Set<TopicPartition> assignment = singleton(new TopicPartition("assignment", 0));
+        Mockito.verify(mockitoConsumer).resume(assignment);
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java]: I do not think we need this verification. Method `resume()` is called in some setup code of the test.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java [hunk 6]
```diff
@@ -2332,19 +2335,12 @@ public void shouldCloseActiveUnassignedSuspendedTasksWhenClosingRevokedTasks() {
...
+        final Set<TopicPartition> assignment = singleton(new TopicPartition("assignment", 0));
+        when(mockitoConsumer.assignment()).thenReturn(assignment);
 
-        // second `handleAssignment`
-        consumer.commitSync(offsets);
-        expectLastCall();
+        when(activeTaskCreator.createTasks(any(), Mockito.eq(taskId00Assignment))).thenReturn(singletonList(task00));
 
-        replay(consumer);
+        taskManager.setMainConsumer(mockitoConsumer);
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:2338]: That could be set in `setupTaskManager()`.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java [hunk 7]
```diff
@@ -2332,19 +2335,12 @@ public void shouldCloseActiveUnassignedSuspendedTasksWhenClosingRevokedTasks() {
...
         // first `handleAssignment`
-        expectRestoreToBeCompleted(consumer);
-        when(activeTaskCreator.createTasks(any(), Mockito.eq(taskId00Assignment))).thenReturn(singletonList(task00));
-        expectLastCall();
-
-        // `handleRevocation`
-        consumer.commitSync(offsets);
-        expectLastCall();
+        final Set<TopicPartition> assignment = singleton(new TopicPartition("assignment", 0));
+        when(mockitoConsumer.assignment()).thenReturn(assignment);
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java]: I do not think you need this in this test. This call is not relevant for this test.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java (hunk 1 above)
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java]: I do not think you need this in this test. This call is not relevant for this test.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java [hunk 8]
```diff
@@ -2369,12 +2366,9 @@ public void closeClean() {
             }
         };
 
-        // first `handleAssignment`
-        expectRestoreToBeCompleted(consumer);
         when(activeTaskCreator.createTasks(any(), Mockito.eq(taskId00Assignment))).thenReturn(singletonList(task00));
-        expectLastCall();
 
-        replay(consumer);
+        taskManager.setMainConsumer(mockitoConsumer);
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:2365]: Set in `setupTaskManager()`.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java [hunk 9]
```diff
@@ -2399,7 +2393,8 @@ public void shouldCloseActiveTasksWhenHandlingLostTasks() throws Exception {
         final StateMachineTask task01 = new StateMachineTask(taskId01, taskId01Partitions, false, stateManager);
 
         // `handleAssignment`
-        expectRestoreToBeCompleted(consumer);
+        final Set<TopicPartition> assignment = singleton(new TopicPartition("assignment", 0));
+        when(mockitoConsumer.assignment()).thenReturn(assignment);
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java]: Not needed because not relevant for this test.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java [hunk 10]
```diff
@@ -2437,6 +2432,7 @@ public void shouldCloseActiveTasksWhenHandlingLostTasks() throws Exception {
 
         assertThat(taskManager.lockedTaskDirectories(), is(emptySet()));
         Mockito.verify(activeTaskCreator).closeAndRemoveTaskProducerIfNeeded(taskId00);
+        Mockito.verify(mockitoConsumer).resume(assignment);
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java]: Not needed because not relevant for this test.

**Review by cadonna — COMMENTED** [review]
Hi @clolov ! Thanks for the PR!

I made a first pass and left a couple of comment that need to be applied to the whole PR.

**Comment by clolov** [comment]
Thank you, I will aim to address the comments as soon as possible!

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java (hunk 3 above)
Comment (clolov) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:4813]: I can try to completely get rid of the `consumer` in this pull request. I wanted to minimise the changes needed to review this change (fix all tests which relied on `expectRestoreToBeCompleted`). I was imagining one more pull request after this one is merged which gets rid of the the remaining usage of `consumer` and moves this into the setup. If you want to see all changes in this pull request I can try to achieve them in a subsequent commit.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java (hunk 6 above)
Comment (clolov) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:2338]: Same reply as in https://github.com/apache/kafka/pull/15261#discussion_r1481555805

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java (hunk 8 above)
Comment (clolov) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:2365]: Same reply as in https://github.com/apache/kafka/pull/15261#discussion_r1481555805

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java (hunk 5 above)
Comment (clolov) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java]: I was trying to keep the same strength of verification as was already present from the method I was trying to get rid of i.e.
```
    private static void expectRestoreToBeCompleted(final Consumer<byte[], byte[]> consumer) {
        final Set<TopicPartition> assignment = singleton(new TopicPartition("assignment", 0));
        expect(consumer.assignment()).andReturn(assignment);
        consumer.resume(assignment); <-- THIS
        expectLastCall();
    }
```
If you think I can relax this, I am happy to remove it from the tests where you have made the same remark

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java (hunk 7 above)
Comment (clolov) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java]: Same reply as in https://github.com/apache/kafka/pull/15261#discussion_r1481560896

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java (hunk 1 above)
Comment (clolov) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java]: Same reply as in https://github.com/apache/kafka/pull/15261#discussion_r1481560896

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java (hunk 9 above)
Comment (clolov) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java]: Same reply as in https://github.com/apache/kafka/pull/15261#discussion_r1481560896

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java (hunk 10 above)
Comment (clolov) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java]: Same reply as in https://github.com/apache/kafka/pull/15261#discussion_r1481560896

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java (hunk 4 above)
Comment (clolov) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java]: I am not certain I follow, do you just want `singleton(new TopicPartition("assignment", 0));` to be encapsulated in its own method?

**Comment by clolov** [comment]
Heya @cadonna, apologies for the delay. I am not certain I fully understand the comments, so I wanted to confirm before making changes. The purpose of this pull request is to just migrate the tests depending on calls to `expectRestoreToBeCompleted`. That method included both a `when` and a `verify` step. This is why when we move it to Mockito-world it becomes a bit more verbose. I understand the `resume` verification might not be the purpose of the test, but it was verified in EasyMock-world and I didn't want to make the assertions more lenient. I have one more pull request after this one which should clean up all these intermediate steps and get rid of the `mockitoConsumer`, that's why I didn't make the changes in the setup method.
If you are happy to review the final changes directly I could try to merge the subsequent pull request into this one.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java (hunk 4 above)
Comment (cadonna) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java]: I see `singleton(new TopicPartition("assignment", 0))` added around 45 times in this PR. I assume this is because `handleAssignment()` uses it internally. I propose to define a field final static variable `assignment` or similar that can be reused. Alternatively, the assignment can also be passed to `handleAssignment()`. Maybe to limit the changes in this PR, you should go for the former option of the field variable.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java (hunk 5 above)
Comment (cadonna) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java]: With EasyMock you need to call `verify()` to verify a call. I do not see a call to `verify()` in the old code. That tells me that verifying the call to `consumer.resume()` was not relevant in the old code of this test. Additionally, the goal of this test, is to verify the result of `taskManager.getTaskOffsetSums()`. For that you need some tasks assigned. So part of the setup of this test is to install an assignment so that you get tasks assigned. With `Mockito.verify(mockitoConsumer).resume(assignment)` you would test rather an unimportant part of the setup of this test.
To sum up, I have the impression that you verify more in the migration that it was verified in the old code and you added a rather unimportant verification.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java (hunk 6 above)
Comment (cadonna) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:2338]: Fair enough! Waiting for your follow-up PR then.

**Review by cadonna — COMMENTED** [review]
Thanks for your replies and patience, @clolov !

Here my replies.

**Comment by clolov** [comment]
Thanks for the review @cadonna! I will provide an updated version tomorrow morning!

**Comment by clolov** [comment]
Heya @cadonna, apologies for the delay! Hopefully the latest version addresses your comments 😊, but if not let me know and I will pick it up tomorrow

**Review by cadonna — APPROVED** [review]
@clolov Thanks for the updates!

LGTM!

**Comment by cadonna** [comment]
test failures are unrelated
